garage/tf/models/base.py

```diff
@@ -0,0 +1,313 @@
+        return self._outputs
+
+    @property
+    def parameters(self):
+        """Parameters of the Model."""
+        raise NotImplementedError
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set model parameters."""
+        raise NotImplementedError
+
+    def __getstate__(self):
+        """Object.__getstate__."""
+        args, kwargs = self.init_spec
+        return {'args': args, 'kwargs': kwargs, 'parameters': self.parameters}
+
+    def __setstate__(self, data):
+        """Object.__setstate__."""
+        # TODO(krzentner): There has got to be a better way.
+        out = type(self)(*data['args'], **data['kwargs'])
+        out._default_parameters = data['parameters']
+        self.__dict__.update(out.__dict__)
+
+
+class TfModel(AutoPickableModel):
+    r"""
+    Model class for TensorFlow.
+
+    This class is a parent-child design.
+    The parent model contains the parameters, i.e. the structure, and
+    two child model are constructed with input_1 and input_2. Inputs and
+    outputs of the parent model will be binded with the default child, which
+    is the first child constructed.
+
+    When new child model is created, it reuses the parameter from the
+    parent model and can be accessed by calling model.child_2, where
+    child_2 is the name given when building this particular child model.
+    If a child is constructed without given a name, the name "default" will
+    be used.
+
+    The design is illustrated as the following:
+
+         input_1                      input_2
+            |                            |
+    ============== Parent (TfModel)===============
+    |       |                            |       |
+    |       |          Parameters        |       |
+    |    ===========  /           \  =========== |
+    |    | Child 1 | /             \ | Child 2 | |
+    |    |(TfModel)|/               \|(TfModel)| |
+    |    ===========                 =========== |
+    |       |                            |       |
+    ==============================================
+            |                            |
+       (model.outputs) OR                |
+    (model.default.outputs)        model.child_2.outputs
+
+    The parameters inside a model will be initialized when calling build().
+    Do not call tf.global_variable_initializers() after building a model as it
+    will reassign random weights to the model.
+
+    Examples are also available in tests/garage/tf/models/test_new_model.
+
+    Args:
+      name (str): Name of the model. It will also become the variable scope
+      of the model. Every model should have a unique name.
+    """
+
+    def __init__(self, name):
+        super().__init__()
+        self._name = name or self._name
+        self._is_parent = True
+        self._is_built = False
+        self._variable_scope = tf.variable_scope(self.name, reuse=False)
+        self._name_scope = tf.name_scope("default")
+        self._siblings = {}
```

Comment:
are they siblings or are they children?

i think siblings/twins/clones is a more apt description. children implies a tree structure (and you will want to reserve that for referrring to hierarchical mdoels). you should make this consistent with the documentation above.

Reply:
I will change the naming accordingly.

Reply:
is the resolution to call this `_graphs` or `_networks` instead then?

Reply:
We can separate into two classes: `Model` and `Networks`, and then have this called `_networks`.

Reply:
okay, that sounds good.

Reply:
i will resolve when you land the Network class